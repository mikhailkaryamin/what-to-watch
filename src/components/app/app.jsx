import React from 'react';
import Main from '../main/main.jsx';
import {arrayOf} from 'prop-types';
import {filmCardPropTypes} from '../../types.js';

const App = (props) => {
  const {filmCards} = props;

  return <Main
    filmCards={filmCards}
  />;
};

App.propTypes = {
  filmCards: arrayOf(
      filmCardPropTypes
  ),
};

export default App;
