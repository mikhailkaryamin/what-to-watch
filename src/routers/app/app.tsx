import * as React from 'react';
import {
  BrowserRouter as Router,
  Route,
  RouteComponentProps,
  Switch,
} from 'react-router-dom';

import {
  CommentRAWType,
  FilmType,
  UserRAWType,
} from '../../types';

import {
  AppRoute,
  AuthStatus,
  NoAvailableMessage,
} from '../../const';

import WithFilm from '../with-film/with-film';
import Private from '../private/private';

import Comment from '../../components/comment/comment';
import Favorites from '../../components/favorites/favorites';
import Main from '../../components/main/main';
import NoAvailable from '../../components/no-available/no-available';
import SignIn from '../../components/sign-in/sign-in';

import withAuthorization from '../../hocs/with-authorization/with-authorization';
import withComment from '../../hocs/with-comment/with-comment';

const CommentWrapped = withComment(Comment);
const SignInWrapped = withAuthorization(SignIn);

type Props = {
  authStatus: string | null;
  films: FilmType[];
  signIn: (userData: UserRAWType) => void;
  statusLoadFilms: string | null;
  statusUploadComment: string | null;
  uploadComment: (commentData: CommentRAWType, id: number) => void;
}

const App: React.FC<Props> = (props: Props) => {
  const {
    authStatus,
    films,
    signIn,
    statusLoadFilms,
    statusUploadComment,
    uploadComment,
  } = props;

  const isAuth = authStatus === AuthStatus.AUTH;
  const isEmptyListFilms = films.length === 0;
  const isLoading = statusLoadFilms === null || authStatus === null;

  return (
    <Router>
      <Switch>
        <Route
          exact
          path={AppRoute.ROOT}
          render={() => {
            return (
              (!isLoading && isEmptyListFilms)
                ? <NoAvailable
                  isAuth={isAuth}
                  isLink={false}
                  isWithSignIn={true}
                  message={NoAvailableMessage.FILMS}
                />
                : <Main />
            );
          }}
        />

        <Route
          exact
          path={AppRoute.LOGIN}
          render={() => {
            return (
              <SignInWrapped
                isAuth={isAuth}
                signIn={signIn}
              />
            );
          }}
        />

        <WithFilm
          exact
          isAuth={isAuth}
          path={AppRoute.PLAYER}
          statusLoadFilms={statusLoadFilms}
        />

        <WithFilm
          exact
          isAuth={isAuth}
          path={AppRoute.FILM}
          statusLoadFilms={statusLoadFilms}
        />

        <Private
          exact={true}
          isAuth={isAuth}
          path={AppRoute.MY_LIST}
          render={() => {
            return (
              <Favorites
                isAuth={isAuth}
              />
            );
          }}
        />

        <Private
          exact={true}
          isAuth={isAuth}
          path={AppRoute.ADD_COMMENT}
          render={(renderProps: RouteComponentProps<any>) => {
            console.log(renderProps)
            const filmId = parseInt(renderProps.match.params.id, 10);

            return (
              <CommentWrapped
                filmId={filmId}
                uploadComment={uploadComment}
                statusUploadComment={statusUploadComment}
              />
            );
          }}
        />

        <Route
          path={AppRoute.NOT_FOUND}
          render={() => {
            return (
              <NoAvailable
                isAuth={isAuth}
                isLink={true}
                isWithSignIn={true}
                message={NoAvailableMessage.PAGE}
              />
            );
          }}
        />

      </Switch>
    </Router>
  );
};

export {App};
