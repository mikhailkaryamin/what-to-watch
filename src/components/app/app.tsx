import * as React from 'react';
import {connect} from 'react-redux';
import {
  CommentRAWType,
  FilmType,
  RootState,
  UserRAWType,
} from '../../types';

import {
  AuthStatus,
  NoAvailableMessage,
  StatusRequestServer,
} from '../../const';

import {Operation as CommentOperation} from '../../reducer/comment/comment';
import {Operation as FavoriteOperation} from '../../reducer/favorite/favorite';
import {Operation as UserOperation} from '../../reducer/user/user';

import {getStatus as getStatusCommentUpload} from '../../reducer/comment/selectors';
import {
  getStatusFilms as getStatusLoadingFilms,
  getStatusPromo as getStatusLoadingPromo,
  getFilms,
} from '../../reducer/films/selectors';
import {getAuthStatus} from '../../reducer/user/selectors';

import Loader from '../loader/loader';
import NoAvailable from '../no-available/no-available';

import {App as Router} from '../../routers/app/app';

type Props = {
  authStatus: string | null;
  films: FilmType[];
  loadFavorite: () => void;
  signIn: (user: UserRAWType) => void;
  statusLoadFilms: string | null;
  statusLoadPromo: string | null;
  statusUploadComment: string | null;
  uploadComment: (commentData: CommentRAWType, id: number) => void;
}

const App: React.FC<Props> = (props: Props) => {
  const {
    authStatus,
    films,
    loadFavorite,
    signIn,
    statusLoadFilms,
    statusLoadPromo,
    statusUploadComment,
    uploadComment,
  } = props;

  const isAuth = authStatus === AuthStatus.AUTH;
  const isErrorNetwork = statusLoadFilms === StatusRequestServer.FAIL;
  const isLoading = statusLoadFilms === null || statusLoadPromo === null || authStatus === null;

  if (isErrorNetwork) {
    return <NoAvailable
      isAuth={isAuth}
      isLink={false}
      isWithSignIn={false}
      message={NoAvailableMessage.ERROR_SERVER}
    />;
  }

  if (isLoading) {
    return (
      <Loader />
    );
  }

  if (isAuth) {
    loadFavorite();
  }

  return (
    <Router
      authStatus={authStatus}
      films={films}
      signIn={signIn}
      statusLoadFilms={statusLoadFilms}
      statusUploadComment={statusUploadComment}
      uploadComment={uploadComment}
    />
  );
};

const mapStateToProps = (state: RootState) => ({
  authStatus: getAuthStatus(state),
  films: getFilms(state),
  statusLoadFilms: getStatusLoadingFilms(state),
  statusLoadPromo: getStatusLoadingPromo(state),
  statusUploadComment: getStatusCommentUpload(state),
});

const mapDispatchToProps = (dispatch: Function) => ({
  loadFavorite: () => {
    dispatch(FavoriteOperation.loadFavoriteFilms());
  },
  signIn: (user: UserRAWType) => {
    dispatch(UserOperation.signIn(user));
  },
  uploadComment: (commentData: CommentRAWType, id: number) => {
    dispatch(CommentOperation.uploadComment(commentData, id));
  },
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
