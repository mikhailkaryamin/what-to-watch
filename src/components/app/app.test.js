import * as React from 'react';
import renderer from 'react-test-renderer';

import {App} from './app.tsx';

describe(`Render App`, () => {

  test(`should render app`, () => {
    const wrapper = renderer
      .create(
          <App
            authStatus={`Auth`}
            films={[]}
            loadFavorite={()=> {}}
            signIn={()=>{}}
            statusLoadFilms={`success`}
            statusUploadComment={`success`}
            uploadComment={() => {}}
          />
      ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});

