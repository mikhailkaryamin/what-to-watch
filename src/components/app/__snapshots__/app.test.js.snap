// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App should render app 1`] = `
Array [
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/aviator.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link "
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <img
            alt="User avatar"
            height="63"
            src="img/avatar.jpg"
            width="63"
          />
        </div>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Devin Albert"
            height="327"
            src="img/aviator.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Devin Albert
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              2015
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn movie-card__button btn--list"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item"
          onClick={[Function]}
        >
          <a
            className="catalog__genres-link"
            href="#"
          >
            Comedy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list "
      />
      
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;

exports[`Render App should render app with open film 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
          src="img/aviator.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link "
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="img/avatar.jpg"
              width="63"
            />
          </div>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Devin Albert
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              2015
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn movie-card__button btn--list"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="add-review.html"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Devin Albert"
            height="327"
            src="img/aviator.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
                onClick={[Function]}
              >
                <a
                  className="movie-nav__link"
                  href="#"
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item "
                onClick={[Function]}
              >
                <a
                  className="movie-nav__link"
                  href="#"
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item "
                onClick={[Function]}
              >
                <a
                  className="movie-nav__link"
                  href="#"
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              3
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Normal
              </span>
              <span
                className="movie-rating__count"
              >
                300
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
      Aliquam a justo elit. Nulla vitae hendrerit dolor. Sed luctus massa lectus.
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Arthur Agee
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Arthur Agee
                Robert Agnew
                Alan Aisenberg
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list catalog--like-this"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;

exports[`Render App should render app with watched film 1`] = `
<div
  className="player"
>
  <video
    className="player__video"
    controls={false}
    poster="img/aviator.jpg"
    src="some/link"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          value={0}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        00:00:00
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Transpotting
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
